tsl/src/chunk.c

```diff
@@ -507,3 +508,187 @@ chunk_drop_replica(PG_FUNCTION_ARGS)
+		ts_hypertable_has_chunks(table_relid, AccessShareLock);
+
+		/* collect a list of chunks which should exists on the specified data node
+		 * according to the access node */
+		chunk_id_list = ts_chunk_get_chunk_ids_by_hypertable_id(ht->fd.id);
+
+		appendStringInfo(cmd, "array[");
+		foreach (lc, chunk_id_list)
+		{
+			int32_t chunk_id = lfirst_int(lc);
+			int32_t node_chunk_id = -1;
+			Chunk *chunk = ts_chunk_get_by_id(chunk_id, true);
+
+			if (ts_chunk_has_data_node(chunk, node_name))
+			{
+				/* we need to use node_chunk_id instead of chunk_id */
+				ScanIterator it = ts_chunk_data_nodes_scan_iterator_create(CurrentMemoryContext);
+				ts_chunk_data_nodes_scan_iterator_set_node_name(&it, node_name);
+
+				/* scan for chunks that reference the given data node */
+				ts_scanner_foreach(&it)
+				{
+					TupleTableSlot *slot = ts_scan_iterator_slot(&it);
+					bool PG_USED_FOR_ASSERTS_ONLY isnull = false;
+					int32_t scan_chunk_id;
+
+					scan_chunk_id =
+						DatumGetInt32(slot_getattr(slot, Anum_chunk_data_node_chunk_id, &isnull));
+					Assert(!isnull);
+
+					if (scan_chunk_id != chunk_id)
+						continue;
+
+					node_chunk_id = DatumGetInt32(
+						slot_getattr(slot, Anum_chunk_data_node_node_chunk_id, &isnull));
+					Assert(!isnull);
+					break;
+				}
+				ts_scan_iterator_close(&it);
+
+				if (!first)
+					appendStringInfo(cmd, ", ");
+
+				appendStringInfo(cmd, "%d", node_chunk_id);
+
+				first = false;
+			}
+		}
+		appendStringInfo(cmd, "]::integer[])");
+
+		/* execute command on the data node */
+		ts_dist_cmd_run_on_data_nodes(cmd->data, list_make1(node_name), true);
+	}
+	else if (membership == DIST_MEMBER_DATA_NODE)
+	{
+		ListCell *lc;
+		List *chunk_id_list;
+
+		if (!hypertable_is_distributed_member(ht))
+			ereport(ERROR,
+					(errcode(ERRCODE_TS_HYPERTABLE_NOT_DISTRIBUTED),
+					 errmsg("hypertable \"%s\" is not distributed member",
+							get_rel_name(table_relid))));
+
+		if (node_name != NULL)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("node_name argument cannot be used on the data node")));
+
+		if (chunks_arr == NULL)
+			ereport(ERROR,
+					(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+					 errmsg("chunks argument cannot be NULL")));
+
+		/* compare each chunk with the chunks from the array */
+		chunk_id_list = ts_chunk_get_chunk_ids_by_hypertable_id(ht->fd.id);
+
+		foreach (lc, chunk_id_list)
+		{
+			int chunk_id = lfirst_int(lc);
+			Chunk *chunk = ts_chunk_get_by_id(chunk_id, true);
+			ArrayIterator it;
+			Datum chunk_id_datum;
+			bool isnull;
+			bool match = false;
+
+			it = array_create_iterator(chunks_arr, 0, NULL);
+
+			while (array_iterate(it, &chunk_id_datum, &isnull))
+			{
+				if (isnull)
+					ereport(ERROR,
+							(errcode(ERRCODE_INVALID_PARAMETER_VALUE),
+							 errmsg("chunks array arguments cannot be NULL")));
+
+				if (chunk_id == DatumGetInt32(chunk_id_datum))
+				{
+					match = true;
```

Comment:
I suggesty putting converting the array into a hashtable before the loop so that you can do O(1) lookups to filter chunks.